ompi/mca/pml/ob1/pml_ob1_accelerator.c

```diff
@@ -0,0 +1,599 @@
+       array. */
+    accelerator_event_htod_frag_array = (struct mca_btl_base_descriptor_t **) malloc(
+        sizeof(struct mca_btl_base_descriptor_t *) * accelerator_event_max);
+    if (NULL == accelerator_event_htod_frag_array) {
+        opal_output_verbose(1, mca_pml_ob1_output, "No memory.");
+        rc = OPAL_ERROR;
+        goto cleanup_and_error;
+    }
+
+cleanup_and_error:
+    if (OPAL_SUCCESS != rc) {
+        if (NULL != accelerator_event_dtoh_array) {
+            free(accelerator_event_dtoh_array);
+        }
+        if (NULL != accelerator_event_dtoh_frag_array) {
+            free(accelerator_event_dtoh_frag_array);
+        }
+        if (NULL != accelerator_event_htod_array) {
+            free(accelerator_event_htod_array);
+        }
+        if (NULL != accelerator_event_htod_frag_array) {
+            free(accelerator_event_htod_frag_array);
+        }
+    }
+
+    return rc;
+}
+
+void mca_pml_ob1_accelerator_fini(void)
+{
+    int i;
+
+    if (0 == strcmp(accelerator_base_selected_component.base_version.mca_component_name, "null")) {
+        return;
+    }
+
+    if (NULL != accelerator_event_htod_array) {
+        for (i = 0; i < accelerator_event_max; i++) {
+            if (NULL != &accelerator_event_htod_array[i]) {
+                opal_accelerator.destroy_event(&accelerator_event_htod_array[i]);
+            }
+        }
+        free(accelerator_event_htod_array);
+    }
+
+    if (NULL != accelerator_event_dtoh_array) {
+        for (i = 0; i < accelerator_event_max; i++) {
+            if (NULL != &accelerator_event_dtoh_array[i]) {
+                opal_accelerator.destroy_event(&accelerator_event_dtoh_array[i]);
+            }
+        }
+        free(accelerator_event_dtoh_array);
+    }
+    if (NULL != accelerator_event_dtoh_frag_array) {
+        free(accelerator_event_dtoh_frag_array);
+    }
+    if (NULL != accelerator_event_htod_frag_array) {
+        free(accelerator_event_htod_frag_array);
+    }
+
+    opal_accelerator.destroy_stream(&htod_stream);
+    opal_accelerator.destroy_stream(&dtoh_stream);
+
+    OBJ_DESTRUCT(&pml_ob1_accelerator_htod_lock);
+    OBJ_DESTRUCT(&pml_ob1_accelerator_dtoh_lock);
+}
+
+size_t mca_pml_ob1_rdma_cuda_btls(
+    mca_bml_base_endpoint_t* bml_endpoint,
+    unsigned char* base,
+    size_t size,
+    mca_pml_ob1_com_btl_t* rdma_btls);
+
+int mca_pml_ob1_accelerator_need_buffers(void * rreq,
+                                         mca_btl_base_module_t* btl);
+
+void mca_pml_ob1_accelerator_add_ipc_support(struct mca_btl_base_module_t* btl, int32_t flags,
+                                             ompi_proc_t* errproc, char* btlinfo);
+
+/**
+ * Handle the accelerator buffer.
+ */
+int mca_pml_ob1_send_request_start_accelerator(mca_pml_ob1_send_request_t* sendreq,
+                                               mca_bml_base_btl_t* bml_btl,
+                                               size_t size) {
+    int rc;
+#if OPAL_CUDA_GDR_SUPPORT
```

Comment:
This ifdef region is not CUDA-specific right? It should probably apply to all accelerator instances.

Reply:
I think this path would need to be tested with Rocm, I'm not exactly sure if this is safe to take out so I left it in

Reply:
As follow up work, will need to go test this with other accelerators, leaving for now.